Implement migration manager

drEfrati opened the pull request:
#### Main purpose:
 To implement the migration manager phase

#### Technical description:
Implement start migration 
Implement kin version check
Implement validate burn check
Implement the burn itself
implement the migrate account 
Implement get current sdk (Kin2/3) 
Implement callbacks for client to know the migration status
implement retry mechanism(not yet in its final version)
implement error handling
* still working on the BI but in a different branch

#### Dilemmas you faced with?
Which came first, the chicken or the egg

#### Tests
Add only integration tests but not yet finished

#### Documentation (Source/readme.md)
need to add a draft and then Sahsa to make the final version

yosriz commented on migration/build.gradle:
```diff
@@ -39,7 +39,7 @@ dependencies {
     implementation supportDependencies.supportAnnotations
 
     implementation 'com.github.kinecosystem:kin-sdk-android:3b3f0c0dfe'
```
we should publish numbered version, and update both sdks to latest (to include new crypto lib version fixes)

yosriz commented on migration/src/androidTest/java/kin/sdk/migration/WhitelistServiceForTest.java:
```diff
@@ -23,7 +24,7 @@
 public class WhitelistServiceForTest implements IWhitelistService {
 
-    private static final String URL_WHITELISTING_SERVICE = "http://10.4.59.1:3003/whitelist";
+    private static final String URL_WHITELISTING_SERVICE = "http://34.239.111.38:3000/whitelist";
```
don't we have a playground whitelister yet?

drEfrati replied:
not yet as far as i know

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+    private val LINE_FULL_RESULT_CODE = "op_line_full"
+
+    private val timeoutDurationSecondsLong: Long = 100 //TODO need to change to a normal number, maybe 20 seconds
+    private val timeoutDurationSecondsShort: Long = 50 //TODO need to change to a normal number, maybe 10 seconds
```
don't you have those timeouts already on sdk/core tests?

drEfrati replied:
Yep, but not set to so long

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+                fail("not supposed to reach onError with this exception: $e")
+            }
+        })
+        latch.await(timeoutDurationSecondsLong, TimeUnit.SECONDS)
```
timeout here should be very low (because it's old kin)

drEfrati replied:
Correct and will be fixed

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+        val newKinClient = getKinClientOnNewKinBlockchain()
+        val newAccount = newKinClient?.getAccount(newKinClient.accountCount - 1)
+        assertEquals(newAccount?.getKinSdkVersion(), KinSdkVersion.NEW_KIN_SDK)
+        latch.countDown()
```
no need for latches here, everything is sync in this level (the rest is taking care inside the helper methods)

drEfrati replied:
correct, I changed it couple of times and didn't noticed

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+            fakeKinIssuer.createAccount(account?.publicAddress.orEmpty())
+        }).start()
+
+        var listenerRegistration: IListenerRegistration? = null
```
why do you need to wait for account creation? `fakeKinIssuer.createAccount` is a blocking call and account should be created if no exception is throwed

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+        val newKinClient = getKinClientOnNewKinBlockchain()
+        val newAccount = newKinClient?.getAccount(newKinClient.accountCount - 1)
+        assertEquals(newAccount?.getKinSdkVersion(), KinSdkVersion.NEW_KIN_SDK)
+        latch.countDown()
```
no need for latches here, everything is sync in this level (the rest is taking care inside the helper methods)

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+    @Test
+    @LargeTest
+    fun start_StartWhileInProcess_MigrationInProcessException() {
+        assertFailsWith(MigrationInProcessException::class) {
```
nice! so you don't need the `expectedEx`...

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+    // Create account on the blockchain and when creation complete then the method will end.
+    private fun createAccount(account: IKinAccount?) {
+        val latch = CountDownLatch(1)
+        // run it on a background thread because this method is called from the main ui thread.
```
same comment as in `activateAccount`

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+            val latch = CountDownLatch(1)
+            val migrationManager = getNewMigrationManager(IKinVersionProvider {
+                // waiting so it will be in the middle of the migration process while starting another one.
+                latch.await(timeoutDurationSecondsShort, TimeUnit.SECONDS)
```
should be the opposite, `countDown` here, and then `await()` outside, in this way you won't need the timeout (and actually now the latch is unnecessary)

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+            }
+
+            override fun onReady(kinClient: IKinClient) {
+                assertTrue(migrationStarted)
```
asserting not on the main test thread is discouraged and can lead to problems. so you should move it to after the `latch.await`

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+    @LargeTest
+    fun start_MigrationAlreadyCompleted_getNewKinClient() {
+        val latch = CountDownLatch(1)
+        getNewKinClientAfterMigration()// blocking call
```
unnecessary comment

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+            }
+
+            override fun onError(e: Exception) {
+                fail("not supposed to reach onError with this exception: $e")
```
asserting not on the main test thread is discouraged and can lead to problems. so you should move it to after the `latch.await`

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+        migrationManager.start(object : MigrationManagerListener {
+
+            override fun onMigrationStart() {
+                fail("not supposed to reach onMigrationStart")
```
same comment calling fail() from a non test thread.

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+            }
+
+            override fun onError(e: Exception) {
+                fail("not supposed to reach onError with this exception: $e")
```
move to main test thread

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+        val migrationManager = getNewMigrationManager(IKinVersionProvider {throw FailedToResolveSdkVersionException()})
+        migrationManager.start(object : MigrationManagerListener {
+            override fun onMigrationStart() {
+                fail("not supposed to reach onMigrationStart")
```
move fails/asserts to main test thread

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+    fun start_AlreadyMigratedButNotSavedLocally_ZeroBalance_AccountNotActivated_getNewKinClient() {
+        val latch = CountDownLatch(1)
+        getNewKinClientAfterMigration()// blocking call
+        teardown()
```
rename it to a meaningful name (clear migration flag?)

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+        val latch = CountDownLatch(1)
+        getNewKinClientAfterMigration()// blocking call
+        teardown()
+        fail("remove this fail after finished with the TODO")//TODO  add funding with faucet
```
implement faucet call

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+            }
+
+            override fun onReady(kinClient: IKinClient) {
+                assertTrue(migrationStarted)
```
move fails/asserts to main test thread

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+            }
+
+            override fun onReady(kinClient: IKinClient) {
+                assertTrue(migrationStarted)
```
move fails/asserts to main test thread

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+                }
+
+                override fun onReady(kinClient: IKinClient) {
+                    assertTrue(migrationStarted)
```
move fails/asserts to main test thread

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+    @Test
+    @LargeTest
+    fun start_AlreadyBurnedAccountButNotMigrated_WithBalance_MigrateAndGetNewKinClient() {
+        //TODO need to implement after we check how to fund the account with the current issuer and the random issuer because then migration could not succeed
```
implement before merge

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+        //TODO need to implement after we check how to fund the account with the current issuer and the random issuer because then migration could not succeed
+    }
+
+    //TODO should we test all kind of scenarios like it was burned but then network or something fail and he didn't migrate
```
we should test of course all possible failures, you will need to have mocks for mocking errors and probably have it as unit test?

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+            }
+
+            override fun onReady(kinClient: IKinClient) {
+                assertTrue(startMigration)
```
move fails/asserts to main test thread

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+
+    // Activate account on the blockchain and when activation is complete then the method will end.
+    private fun activateAccount(account: IKinAccount?) {
+        val latch = CountDownLatch(1)
```
are you sure? tests thread is not the main one, and there shouldn't be a problem to perform network request, unless you are calling it from migration callback that triggered on main thread? anyway I don't see such a case here

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+            }
+
+            override fun onError(e: Exception) {
+                fail("not supposed to reach onError with this exception: $e")
```
move fails/asserts to main test thread

yosriz commented on migration/src/androidTest/kotlin/kin/sdk/migration/MigrationIntegrationTest.kt:
```diff
@@ -0,0 +1,428 @@
+            }
+
+            override fun onError(e: Exception) {
+                fail("not supposed to reach onError with this exception: $e")
```
move fails/asserts to main test thread

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
-        return kinClient;
-    }
+	private static final String TAG = MigrationManager.class.getSimpleName();
+	private final static String MIGRATION_MODULE_PREFERENCE_FILE_KEY = "MIGRATION_MODULE_PREFERENCE_FILE_KEY";
```
add KIN as prefix so it will be clearer

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
-        }
-        return kinClient;
-    }
+	private static final String TAG = MigrationManager.class.getSimpleName();
```
OCD - some are `static final` and some are `final static`

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+	private String storeKey;
+	private Handler handler;
+	private OkHttpClient okHttpClient;
+	private volatile AtomicBoolean inMigrationProcess; // defence against multiple calls
```
no need for `volatile` in `AtomicBoolean` (as it's implemented in synchronized fashion)

drEfrati replied:
Correct, start with a volatile boolean and when changed to AtomicBoolean then didn't removed the volatile keyword.

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+	private final static int MAX_RETRIES = 3;
+	private static final String URL_MIGRATE_ACCOUNT_SERVICE = "https://migration-devplatform-playground.developers.kinecosystem.com/migrate?address=";
+
+	private Context context;
```
almost everything here can be final isn't it?

drEfrati replied:
Yep

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+	public MigrationManager(@NonNull Context context, @NonNull String appId,
+		@NonNull MigrationNetworkInfo migrationNetworkInfo,
+		@NonNull IKinVersionProvider kinVersionProvider, @NonNull String storeKey) {
+		this.context = context;
```
save `applicationContext` for being on the safe side

drEfrati replied:
I have no control over what context the developer will give us(although there is a code which can validate if a context is the application context).
I can change the name and the documentation so it will be clear for the developer.
What do you think?

drEfrati replied:
fixed

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+	private String storeKey;
+	private Handler handler;
+	private OkHttpClient okHttpClient;
+	private volatile AtomicBoolean inMigrationProcess; // defence against multiple calls
```
isMigrationInProcess?

drEfrati replied:
yep...

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+				}
+			}).start();
+		} else {
+			inMigrationProcess.set(false);
```
shouldn't set to false here, you will override a different execution state.

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+			new Thread(new Runnable() {
+				@Override
+				public void run() {
+					Log.d(TAG, "starting the migration process in a background thread");
```
don't log on production/release, I suggest to have a specific log tag, and or a flag for matching with what we have now in ecosystem SDK

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+				if (kinSdkVersion == null) {
+					fireOnError(migrationManagerListener, new FailedToResolveSdkVersionException());
+				}
+				if (kinSdkVersion == KinSdkVersion.NEW_KIN_SDK) {
```
add else or return, o.w. you will continue in this case

drEfrati replied:
Correct, I hate those last minute changes when you about to release and forget to do some obvious stuff...

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+		throws OperationFailedException {
+		KinClientCoreImpl kinClientCore = initKinCore();
+		if (kinClientCore.hasAccount()) {
+			migrationManagerListener.onMigrationStart();
```
should be posted on the main thread like the rest

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+			try {
+				Log.d(TAG, "burnAccount: ");
+				ITransactionId transactionId = account.sendBurnTransactionSync(publicAddress);
+				return transactionId.id() != null && !transactionId.id().isEmpty();
```
throw an explicit error in case of empty id instead of boolean

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+						continue;
+					}
+				}
+				throw new MigrationFailedException(e);
```
add some message here specifying the failed phase ("burning old account failed")

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+				return burnAccount(publicAddress, account);
+			}
+		} else {
+			throw new MigrationFailedException("account not valid");
```
specify that public address is null

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+				if (burnProcessSucceeded) {
+					migrateToNewKin(publicAddress, migrationManagerListener);
+				} else {
+					fireOnError(migrationManagerListener,
```
this is not a clear exception, we can't understand what exactly went wrong, and in general, there are some exceptions that are caught inside the burn logic and some outside (in the parent method), I think it should be clear in every exception in the process what is the exact operation that failed.

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+			} catch (OperationFailedException e) {
+				Throwable cause = e.getCause();
+				// Check if it is an http exception with 5xx code and if yes then retry until max tries reached.
+				if (cause instanceof HttpResponseException) { // TODO: 30/12/2018 maybe need to catch more exception for the retry? like java.net.ConnectException?
```
logic can be shared with the previous method

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+			} catch (OperationFailedException e) {
+				Throwable cause = e.getCause();
+				// Check if it is an http exception with 5xx code and if yes then retry until max tries reached.
+				if (cause instanceof HttpResponseException) { // TODO: 30/12/2018 maybe need to catch more exception for the retry? like java.net.ConnectException?
```
let's discuss what should we retry on

drEfrati replied:
Ok, and also will discuss retry strategy

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+
+	private void migrateToNewKin(final String publicAddress, final MigrationManagerListener migrationManagerListener) {
+		Log.d(TAG, "migrateToNewKin: sending the request to migrate");
+		sendRequest(URL_MIGRATE_ACCOUNT_SERVICE + publicAddress, new Callback() {
```
why we need async method? we're in a thread

drEfrati replied:
We don't, in the first iteration I did it and now not really needed.
Make it now sync

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+				if (code != null) {
+					switch (code) {
+						case "4001":  // account not burned
+							exception = new MigrationFailedException(message);
```
specify the phase and the error (account not burned)

drEfrati replied:
It is written in the message we got from the server

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+						case "4003":  // public address is not valid, meaning the format of it is not valid.
+							exception = new MigrationFailedException(message);
+							break;
+						case "4041":  // account was not found
```
how come account is not found? and why do we assume it's ok? :thinking:

drEfrati replied:
I talked with Ron about it and it is probably won't happen but if it will then currently we are just handling it just like a regular account not found, meaning there is no account so we will just give you the new kin sdk.
We can talked with Ron about it again.

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+							handled = true;
+							break;
+						default:
+							exception = new MigrationFailedException();
```
add code and message to exception in case we're not handling it

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+					}
+				}
+			} catch (IOException e) {
+				exception = new MigrationFailedException(e);
```
again be more clear, why can it happen? (Json parsing? network error?)

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+				exception = new MigrationFailedException(e);
+			}
+		} else {
+			exception = new MigrationFailedException();
```
body is null, worth adding the response code

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+			@Override
+			public void run() {
+				if (migrationManagerListener != null) {
+					inMigrationProcess.set(false);
```
toggle the flag outside, also null `migrationManagerListener` shouldn't prevent setting this flag

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+			@Override
+			public void run() {
+				if (migrationManagerListener != null) {
+					if (needToSave) {
```
null migrationManagerListener shouldn't prevent setting the flag and saving, also can be done outside

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationManager.java:
```diff
@@ -1,60 +1,404 @@
+			Log.d(TAG, "Retrying new request");
+			Request newRequest = req.newBuilder().build();
+			Response another = chain.proceed(newRequest);
+			while (retryCounter < MAX_RETRIES && another.code() >= 500) {
```
why we need recursion? I think it makes the code difficult to follow, also there's a subtle bug here, you will return the first retry response (`another`), even if it failed and will be retried. (probably need to add `return` before the recursion `retryRequest`.
additionally, retry counter is a field of this class that is shared among several requests, while everything should happen sequentially here, it's not a safe code.

drEfrati replied:
The field is only shared in the interceptor class and is getting set to 1 when finished.
Although after you mentioned it i start testing(wanted to test it only after Tuesday) and it seems that i need to set it to one only after finished with the intercept method and not in the retry itself.
Regarding the retry itself:
The method is recursive because it was easy for me to think recursively in this case
Second, only if the response was 500 and bigger than a retry will happen.
First there will be an immediate call(a retry) and then starting a loop of retries.
For example if we want to have 3 retries then the first one will occur and then w while loop which will cause to another max of 2 retries.
If you still don't understand or think there is an error then let us do a review for this method next to each other.

yosriz replied:
The recursion is the least problem here, I don't mind keeping it. 
I still think the retry should be var in the method for safety, and still, you will return only first response and ignore the retry response, which means retry will be worthless.

drEfrati replied:
We will talked about it when we will meet.

yosriz commented on migration/src/main/java/kin/sdk/migration/MigrationNetworkInfo.java:
```diff
@@ -0,0 +1,41 @@
+
+public class MigrationNetworkInfo {
+
+    private String coreNetworkUrl;
```
all the fields can be final

yosriz commented on migration/src/main/java/kin/sdk/migration/core_related/KinAccountCoreImpl.java:
```diff
@@ -82,6 +91,48 @@ public ITransactionId sendTransactionSync(@NonNull String publicAddress, @NonNul
     }
 
+    private String addAppIdToMemo(@Nullable String memo, @NonNull String appId) {
+        if (memo != null) {
```
appId should be added even if the memo is null/empty. you can copy the impl from kin-sdk

yosriz commented on migration/src/main/java/kin/sdk/migration/core_related/KinAccountCoreImpl.java:
```diff
@@ -23,13 +24,19 @@
 
 public class KinAccountCoreImpl implements IKinAccount {
 
+    private static String MEMO_APP_ID_VERSION_PREFIX = "1";
```
add final to both

yosriz commented on migration/src/main/java/kin/sdk/migration/core_related/KinClientCoreImpl.java:
```diff
@@ -17,13 +17,15 @@
 public class KinClientCoreImpl implements IKinClient {
 
     private final KinCoreEnvironment kinCoreEnvironment;
+    private String appId;
```
final

yosriz commented on migration/src/main/java/kin/sdk/migration/exception/MigrationFailedException.java:
```diff
@@ -0,0 +1,20 @@
+
+public class MigrationFailedException extends OperationFailedException {
+
+    public MigrationFailedException() {
```
I would leave here just the version with a message and throwable, as every exception should have message and cause (like commented above in `MigrationManager`

yosriz commented on migration/src/main/java/kin/sdk/migration/exception/MigrationInProcessException.java:
```diff
@@ -0,0 +1,16 @@
+
+public class MigrationInProcessException extends OperationFailedException {
+
+    public MigrationInProcessException(Throwable cause) {
```
do we need all the overloads?

drEfrati replied:
I removed all of them but the one who gets only a string and the one who gets a string and a cause, although i can also remove the one who gets only a string and the the invoker can just put null, for me it is more nice this way.

yosriz commented on migration/src/main/java/kin/sdk/migration/interfaces/IKinAccount.java:
```diff
@@ -170,4 +174,9 @@ ITransactionId sendTransactionSync(@NonNull String publicAddress, @NonNull BigDe
     IListenerRegistration addAccountCreationListener(final IEventListener<Void> listener);
 
+    /**
+     * @return true if current account is in the new kin blockchain.,
```
comment is not updated

yosriz commented on migration/src/main/java/kin/sdk/migration/interfaces/IKinVersionProvider.java:
```diff
@@ -1,7 +1,10 @@
 public interface IKinVersionProvider {
 
-    boolean isKinSdkVersion();
+    KinSdkVersion getKinSdkVersion(String appId) throws FailedToResolveSdkVersionException;
```
why do we need appId here?

yosriz commented on migration/src/main/java/kin/sdk/migration/interfaces/MigrationManagerListener.java:
```diff
@@ -0,0 +1,22 @@
+    void onMigrationStart();
+
+    /**
+     * Method is invoked when kinClient is ready to use.
```
whether migration happened or not (final method either this or onError will be fired)

yosriz commented on migration/src/main/java/kin/sdk/migration/interfaces/MigrationManagerListener.java:
```diff
@@ -0,0 +1,22 @@
+    /**
+     * Method is invoked before the migration process itself will start.
+     */
+    void onMigrationStart();
```
specify that's an optional, only will be called when migration takes place

yosriz commented on migration/src/main/java/kin/sdk/migration/interfaces/MigrationManagerListener.java:
```diff
@@ -0,0 +1,22 @@
+    void onReady(IKinClient kinClient);
+
+    /**
+     * Method is invoked when an error occurred in the migration process
```
ends the process

yosriz commented on migration/src/main/java/kin/sdk/migration/sdk_related/KinAccountSdkImpl.java:
```diff
@@ -44,42 +44,48 @@ public String getPublicAddress() {
+                if (whitelistTransactionResult.shouldSendTransaction()) {
+                    transactionId = kinAccount.sendWhitelistTransactionSync(whitelistTransactionResult.getWhitelistedTransaction());
+                }
                 return new KinSdkTransactionId(transactionId);
```
probably a potential bug, in case we're not sending the transaction tx id will be null and can raise issue in the receiver side.

drEfrati replied:
We need to return something so if transaction id is null the developers should handle it.
The developer itself need to expect this because he was the one who decides not to send the transaction

yosriz commented on sample/src/main/java/kin/sdk/migration/sample/ChooseNetworkActivity.java:
```diff
@@ -52,12 +53,28 @@ private void initWidgets() {
+
+            @Override
+            public void onMigrationStart() {
+                // TODO: 02/01/2019 add some progress bar
```
what about all those TODOs?

drEfrati replied:
Will implement it later when we will finish with the important stuff

yosriz commented on sample/src/main/java/kin/sdk/migration/sample/TransactionActivity.java:
```diff
@@ -192,7 +193,7 @@ public void onSuccess(String whitelistTransaction) {
         @Override
         public void onFailure(Exception e) {
-            Utils.logError(e, "whitelistTransaction");
+            Utils.logError(e, "onWhitelistableTransactionReady");
```
WhitelistServiceListener

drEfrati commented:
@yosriz addressed most of the issues except the ones that i wrote to you that we need to discuss about them.
Regarding all the review of the integration tests - i will address it later as i was told not to address it yet.
And at any case i didn't finish with them.
